### ISSUE TITLE
upstream nt36xxx driver

### ISSUE THREAD

**Issue by 99degree** [99degree/linux#6]
the first attempt for upstream is at:

https://lore.kernel.org/lkml/c10b0b99-5bf2-47d3-b879-872f80e6883d@kernel.org/

**Comment by Marc-Pierre-Barbier** [comment]
incredible. if you need testing i will have a lot of freetime by the end of next week.

**Comment by 99degree** [comment]
> incredible. if you need testing i will have a lot

thx and that’s appreciated.

> And i have a working redmi note 9

i am afraid your miatoll use other touch panel that nvt. you should be able to boot into pmos but without touch function. if you interested, plz feel free to hands on.

**Comment by Marc-Pierre-Barbier** [comment]
i build it and installed it using the following commands
```
git clone https://github.com/99degree/linux --depth=1 --branch next
cd linux

wget https://raw.githubusercontent.com/kalimokai/pmbootstrap/refs/heads/offical/helpers/envkernel.sh
source envkernel.sh

git fetch origin config
git branch config FETCH_HEAD
git checkout config -- arch/arm64/configs/sm71xx-xiaomi.config          
# i then edited sm71xx-xiaomi.config to disable the tux logo (to be able to differenciate from the stock miatoll pm kernel)

make defconfig sm71xx-xiaomi.config 
make -j$(nproc) 

pmbootstrap build --envkernel linux-postmarketos-qcom-sm7125
pmbootstrap flasher flash_kernel
fastboot reboot
#penguins are not visible => pmbootstrap did use the right kernel
```

here what i got:
* adb is not working
* usb keyboard not working
* no usb power
* sddm lag (no gpu perhaps ?)
* sleep doesn't turn off the panel
* no tactile

except for the fact i removed the tux logos this feels like exactly the same kernel.

did i make a mistake ? was 99degree/linux:next identical to pmos' kernel ?
what branch should i use ?

**Comment by Marc-Pierre-Barbier** [comment]
i switched to you branch of pmaports and added the packages you added in your ci.

But now i get an error saying that super failed to mount.

i tried reflashing the stock super to no avail.

**Comment by 99degree** [comment]
yes, the build under my account does not proof for ufs pmos. my suggedtion is stick with a working branch of your own.

and you can use official pmbootstrap then merge the firmware miatoll apk source code of pmaport. in additional with adbd-linux. so better start with official build（by pmbootstrap init, in your machine） and add addition apk package such firmware-xiaomi-miatoll and adbd-linux apk. thus you can have both advantage.

or you can install firmware files into your own rootfs/lib/firmware. note that ufs build might have a special root disk img structure and not easily modify this way rather than sdcard build.

my build purposely disable the ufs hw in device tree, and lack support of dynamic partition apk. if you wanna enable this you need to compile your own kernel and add support for dynamic partition. 

if you wanna have adbd, you have to dig with
https://gitlab.com/postmarketOS/pmaports/-/issues/2605

if you have any plan for discuss plz create new issue ticket

happy hacking

**Comment by 99degree** [comment]
according to the mail list, there is a nt36xxx driver already. so if supporting fw download mode, have to enhance the upstream driver by modifying mainline code. most importantly maintain credit of existing driver auther. 

so a re-re-rewrite is needed. and need to base on existing code.

**Comment by 99degree** [comment]
A similar ic from nvt is employed. Some discussion message worth to have a reference. 

https://github.com/map220v/sm8550-mainline/issues/1#issuecomment-2939404101

Quote it here too.
Open
What is the status of this project?
#1
Jump to bottom
Description
@TheMojoMan
TheMojoMan
opened on Oct 21, 2024
First: Thank you so much for your ongoing fantastic work for the Xiaomi Pad 5 and now also for the 6S Pro!

Can you please give a short overview of what is already working on the 6S Pro and what is still missing?
[I would offer my help in testing but I do not own a 6S Pro (yet). I am also not 100% sure if I will buy one but if Linux would work on it, it might clear up my mind.]

Thanks again!!!

Activity
map220v commented on Oct 21, 2024
@map220v
map220v
on Oct 21, 2024
Owner
It currently has working:
UFS,
Type-C,
DisplayPort(2 lane),
Charger,
FuelGauge,
Display(144hz only),
Backlight,
GPU(mesa 24.2 and higher for OpenGL),
WiFi,
Bluetooth

And not working:
Touchscreen(it sends raw frames instead of regular x/y),
Microphones(will fix all of them after touchscreen),
Fingerprint,
Hall sensors(will try after touchscreen fix),
Sensors(untested),
Flash Light & RGB Led(will add after touchscreen),
Speakers(will try fixing after touchscreen, currently i2s clock doesn't seem to work),
Cameras(will try when cams support for sm8550 upstreamed)

TheMojoMan commented on Oct 22, 2024
@TheMojoMan
TheMojoMan
on Oct 22, 2024
Author
That is so impressive!
Already useable as a full computer with keyboard and mouse. And media consumption should be possible as well if we could use bluetooth headphones.

It would be a real shame if Xiaomi really disables the ability to unlock the boot loader as I have recently read about.

alghiffaryfa19 commented on Mar 4
@alghiffaryfa19
alghiffaryfa19
on Mar 4
@map220v what is the current status?

map220v commented on Mar 4
@map220v
map220v
on Mar 4
Owner
Touchscreen now partially works. Currently only one finger is supported and position detection is a bit inaccurate.
All Microphones are working, but in mono mode, so only one microphone can be used at time.
Hall sensor for magnetic cover works.
Flashlight and RGB Led are working.
All Speakers are working.
Analog headset/headphones maybe also works, if type-c audio jack detection is working.
Cameras work in sheng-6.13-camera-wip branch and in future sheng-6.14 branch.
Suspend works now.

UCM2 configs for Speakers and Microphones can be found here https://github.com/map220v/alsa-ucm-conf and firmwares are here https://github.com/map220v/sheng-firmware

alghiffaryfa19 commented on Mar 5
@alghiffaryfa19
alghiffaryfa19
on Mar 5
@map220v xiaomi official keyboard work?

map220v commented on Mar 5
@map220v
map220v
on Mar 5
Owner
@map220v xiaomi official keyboard work?

It doesn't. There is no source code for that driver in sheng branch of Xiaomi repos.

alghiffaryfa19 commented on Mar 5
@alghiffaryfa19
alghiffaryfa19
on Mar 5
Maybe this ?
https://github.com/MiCode/kernel_devicetree/blob/sheng-u-oss/qcom/sheng-sm8550.dtsi
@map220v

alghiffaryfa19 commented on Mar 5
@alghiffaryfa19
alghiffaryfa19
on Mar 5 · edited by alghiffaryfa19
https://github.com/MiCode/kernel_devicetree/blob/c4e85e1569f8acbbfe8e6dd694623aa56e424c60/qcom/sheng-sm8550.dtsi#L760
https://t.me/pipa_mainline/413
pipa-mainline/linux@6301a92
pipa-mainline/linux@8384ccf
https://github.com/MiCode/Xiaomi_Kernel_OpenSource/blob/sheng-u-oss/drivers/power/supply/pogoChg/extSOC/nanosic_platform.c
https://github.com/MiCode/Xiaomi_Kernel_OpenSource/blob/29db5d592c5b577aa9d9611da988a5e071d100dd/arch/arm64/configs/vendor/sheng_GKI.config#L421

do the pipa and sheng have the same use of nanosic 803?

map220v commented on Mar 5
@map220v
map220v
on Mar 5
Owner
Tested nanosic from pipa-mainline with some changes, even with correct pin config and all needed regulators enabled it fails to read nanosic version with "regmap write cmd failed" error. Also on sheng nanosic driver seem to upload nanosic firmware after each boot, so it's probably required for it to work.

alghiffaryfa19 commented on Mar 5
@alghiffaryfa19
alghiffaryfa19
on Mar 5
what about the ones in the sheng device tree that I sent?

alghiffaryfa19 commented on Mar 5
@alghiffaryfa19
alghiffaryfa19
on Mar 5
I want to try adapting it for Sheng, but I haven't succeeded in unlocking the bootloader, HyperOS is quite difficult

map220v commented on Mar 5
@map220v
map220v
on Mar 5 · edited by map220v
Owner
what about the ones in the sheng device tree that I sent?

https://github.com/MiCode/Xiaomi_Kernel_OpenSource/blob/sheng-u-oss/drivers/power/supply/pogoChg/extSOC/nanosic_platform.c is for charging through pogo pin. The one that is for pogo pin keyboard is in vendor_opensource_touch-drivers repo I guess, but Xiaomi didn't shared it.
Edit: sheng's nanosic driver build path is "vendor/xiaomi/proprietary/keyboard/keyboard_driver/sheng/nanosic_driver/", so they didn't upload this driver on purpose.

99degree commented 19 hours ago
@99degree
99degree
19 hours ago
good job man!

if you find touch driver not very functional, you might try below driver:
https://lore.kernel.org/all/20241015-nt36xxx-v1-0-3919d0bffee6@gmail.com/

or a more resent one at
https://github.com/99degree/linux/tree/nt36xxx_stable

alghiffaryfa19 commented 7 hours ago
@alghiffaryfa19
alghiffaryfa19
7 hours ago
good job man!

if you find touch driver not very functional, you might try below driver: https://lore.kernel.org/all/20241015-nt36xxx-v1-0-3919d0bffee6@gmail.com/

or a more resent one at https://github.com/99degree/linux/tree/nt36xxx_stable

support for nt36532e ?

99degree commented 4 hours ago
@99degree
99degree
4 hours ago
good job man!
if you find touch driver not very functional, you might try below driver: https://lore.kernel.org/all/20241015-nt36xxx-v1-0-3919d0bffee6@gmail.com/
or a more resent one at https://github.com/99degree/linux/tree/nt36xxx_stable

support for nt36532e ?

not sure if it need a big code add. i am feeling positive with nt36532e.

my rewrite of nt36xxx driver is based on prev version of nvt driver and this newer nvt driver still compatible with nt36675 and add support for nt36532e so they might shared many common logic.

map220v commented 4 hours ago
@map220v
map220v
4 hours ago
Owner
good job man!

if you find touch driver not very functional, you might try below driver: https://lore.kernel.org/all/20241015-nt36xxx-v1-0-3919d0bffee6@gmail.com/

or a more resent one at https://github.com/99degree/linux/tree/nt36xxx_stable

Thanks, but unfortunately Xiaomi only provides host processing version of Novatek firmware now, so it won't work without rewriting touch handle part.

Also I checked both versions, they look good, but I have few notes:

Novatek 36xxx series TDDI's seem to have only 3 input voltages: VDDI, AVDD, AVEE
Bootloader reset command can cause NDS32 in Novatek chip to crash if DSI part is initializing at that moment
Checks for panel vendors using "bootargs" usually aren't allowed in upstream, it is usually done by multiple dts or by requesting panel vendor id from DSI part(if it was programmed in MTP by panel vendor)
map220v commented 4 hours ago
@map220v
map220v
4 hours ago
Owner
@alghiffaryfa19 I have almost finished nanosic driver, it has hotplug detection and should support both keyboard and touchpad. Check it when you have time.
nanosic.patch

alghiffaryfa19 commented 3 hours ago
@alghiffaryfa19
alghiffaryfa19
3 hours ago
@alghiffaryfa19 I have almost finished nanosic driver, it has hotplug detection and should support both keyboard and touchpad. Check it when you have time. nanosic.patch

Wow, okay i will try it later, thank you for your works

99degree commented 18 minutes ago
@99degree
99degree
18 minutes ago
good job man!
if you find touch driver not very functional, you might try below driver: https://lore.kernel.org/all/20241015-nt36xxx-v1-0-3919d0bffee6@gmail.com/
or a more resent one at https://github.com/99degree/linux/tree/nt36xxx_stable

Thanks, but unfortunately Xiaomi only provides host processing version of Novatek firmware now, so it won't work without rewriting touch handle part.

Yes, feel the same. To be honest, the driver rewrite is pending for long time. So use it works as a kind of mid term solution may be ok but nothing more. the reason and idea for long history tracking for upstream is at 99degree/linux#6

So I am glad if you have any plan to add support for that ic. I am also understand if mod for nvt driver is a fast path too. So enjoy it.

If you decided to mod nt36xxx driver, here are some backgrounds and see if useful for you. Last year i was trying to abstract core logic and split spi and i2c as well as soc variants. it was targeted to support new nt51xxx i2c ic too but give up since out of resources.

Latest development code is in
https://github.com/99degree/linux/tree/nt36xxx_dev

Also I checked both versions, they look good, but I have few notes:

Novatek 36xxx series TDDI's seem to have only 3 input voltages: VDDI, AVDD, AVEE
I tested integrated lcd panel and touch panel, all voltage regulators are being processed in panel driver/DTS node, and touch driver is act only as panel follower, so don't need to handle any resources and pm runtime, except gpio. In other word, those independence touch driver use cases are untested.

Bootloader reset command can cause NDS32 in Novatek chip to crash if DSI part is initializing at that moment
No idea if panel follower helps. I use it as panel follower so every screen on/ pwr button on, fw download is called in. It has been short but still gets into SPI write fail when screen on/off extreme toggling with a warn and no harm. I didn't (too lazy to) test the reset gpio function so there is no answer of hw ic hung. I expect panel regulator shut off then switch on can reset touch ic. And only after panel is ready, then drm subsys issue panel_follower_prepared() for touch driver.

If existing reset logic is incompatible, that might be a more better method to do. One can check if that looks like a kind of new reset like nt51xxx. and it is easy to add a new reset handler like MMAP_ENG_RST_ADDR / MMAP_SW_RST_ADDR or any. so define a new nt36532e_memory_maps may help.

https://github.com/99degree/linux/blob/8335cd53341054eb83928be96e6e4d1f9e644e27/drivers/input/touchscreen/nt36xxx_core.c#L567

Checks for panel vendors using "bootargs" usually aren't allowed in upstream, it is usually done by multiple dts or by requesting panel vendor id from DSI part(if it was programmed in MTP by panel vendor)
yes, currently u-boot forks for sm7125/sm7150 are in a good shape. that checking can be done in uboot with dtb overlay or any method. so far i dont have any vendor id probe logic/info/addr for mipi panel. so this is a possible solution too.

Eventally no need to have such tianma keyword checking thing. Currently code logic is exist but also optional now. The reason of this logic exists is to enable a hack for auto probe detect tianma/huaxing varants for single devicetree DTS file and reduce complexity for PMOS kernel/device apk use years ago.
